src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -305,7 +305,20 @@ impl<'a, 'b: 'a> Context<'a, 'b> {
         }).filter_map(|pkg| {
             pkg.targets().iter().find(|&t| self.is_relevant_target(t))
                .map(|t| (pkg, t))
-        }).collect()
+        }).collect::<Vec<_>>();
+
+        if !target.profile().is_custom_build() &&
```

Comment:
Could you add a comment about the high-level thing that is happening here? 

Is it something like:

> Ensure that the main package is considered a dependency of "subtargets" like tests and examples.